File: src/main/java/net/greenjab/fixedminecraft/registry/ItemRegistry.java
package net.greenjab.fixedminecraft.registry;


import net.greenjab.fixedminecraft.FixedMinecraft;
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookAdditionsComponent;
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookItem;
import net.greenjab.fixedminecraft.registry.item.EchoFruitItem;
import net.greenjab.fixedminecraft.registry.item.TotemItem;
import net.minecraft.block.Block;
import net.minecraft.component.ComponentType;
import net.minecraft.component.DataComponentTypes;
import net.minecraft.component.type.ConsumableComponent;
import net.minecraft.component.type.DeathProtectionComponent;
import net.minecraft.component.type.FireworksComponent;
import net.minecraft.component.type.FoodComponents;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.item.AnimalArmorItem;
import net.minecraft.item.BlockItem;
import net.minecraft.item.FireworkRocketItem;
import net.minecraft.item.Item;
import net.minecraft.item.consume.ApplyEffectsConsumeEffect;
import net.minecraft.item.consume.UseAction;
import net.minecraft.item.equipment.ArmorMaterials;
import net.minecraft.potion.Potion;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.registry.entry.RegistryEntry;
import net.minecraft.sound.SoundEvents;
import net.minecraft.util.Rarity;

import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.UnaryOperator;

public class ItemRegistry {

    public static final Item DRAGON_FIREWORK_ROCKET = register(
            "dragon_firework_rocket", FireworkRocketItem::new, new Item.Settings().component(DataComponentTypes.FIREWORKS, new FireworksComponent(1, List.of()))
    );
    public static final Item MAP_BOOK = register("map_book", MapBookItem::new, new Item.Settings().maxCount(1));
    public static final ComponentType<MapBookAdditionsComponent> MAP_BOOK_ADDITIONS = registerComponent("map_book_additions", (builder) -> builder.codec(MapBookAdditionsComponent.CODEC).packetCodec(MapBookAdditionsComponent.PACKET_CODEC).cache());

    public static final Item NETHERITE_HORSE_ARMOR = register(
            "netherite_horse_armor",
            /* method_63974 */ settings -> new AnimalArmorItem(ArmorMaterials.NETHERITE, AnimalArmorItem.Type.EQUESTRIAN, SoundEvents.ENTITY_HORSE_ARMOR, false, settings),
            new Item.Settings().maxCount(1).fireproof().rarity(Rarity.RARE));

    public static final Item BROKEN_TOTEM = register("broken_totem", Item::new, new Item.Settings().maxCount(1).rarity(Rarity.UNCOMMON));
    public static final Item ECHO_TOTEM = register(
            "echo_totem", TotemItem::new, new Item.Settings().maxCount(1).rarity(Rarity.UNCOMMON).component(DataComponentTypes.DEATH_PROTECTION, DeathProtectionComponent.TOTEM_OF_UNDYING));
    public static final Item ECHO_FRUIT = register(
            "echo_fruit", EchoFruitItem::new, new Item.Settings().maxCount(1).rarity(Rarity.UNCOMMON).food(FoodComponents.CHORUS_FRUIT));

    public static final Item NETHERITE_ANVIL = register(BlockRegistry.NETHERITE_ANVIL, new Item.Settings().fireproof());
    public static final Item CHIPPED_NETHERITE_ANVIL = register(BlockRegistry.CHIPPED_NETHERITE_ANVIL, new Item.Settings().fireproof());
    public static final Item DAMAGED_NETHERITE_ANVIL = register(BlockRegistry.DAMAGED_NETHERITE_ANVIL, new Item.Settings().fireproof());

    public static final Item COPPER_RAIL = register(BlockRegistry.COPPER_RAIL);
    public static final Item EXPOSED_COPPER_RAIL = register(BlockRegistry.EXPOSED_COPPER_RAIL);
    public static final Item WEATHERED_COPPER_RAIL = register(BlockRegistry.WEATHERED_COPPER_RAIL);
    public static final Item OXIDIZED_COPPER_RAIL = register(BlockRegistry.OXIDIZED_COPPER_RAIL);

    public static final Item WAXED_COPPER_RAIL = register(BlockRegistry.WAXED_COPPER_RAIL);
    public static final Item WAXED_EXPOSED_COPPER_RAIL = register(BlockRegistry.WAXED_EXPOSED_COPPER_RAIL);
    public static final Item WAXED_WEATHERED_COPPER_RAIL = register(BlockRegistry.WAXED_WEATHERED_COPPER_RAIL);
    public static final Item WAXED_OXIDIZED_COPPER_RAIL = register(BlockRegistry.WAXED_OXIDIZED_COPPER_RAIL);



    public static final ConsumableComponent GLOW_BERRIES_EFFECT = food()
            .consumeEffect(new ApplyEffectsConsumeEffect(new StatusEffectInstance(StatusEffects.GLOWING, 10, 0), 1F))
            .build();
    public static ConsumableComponent.Builder food() {
        return ConsumableComponent.builder().consumeSeconds(1.6F).useAction(UseAction.EAT).sound(SoundEvents.ENTITY_GENERIC_EAT).consumeParticles(true);
    }

    /** This is used, IntelliJ just doesn't realise */
    public static final RegistryEntry<Potion> BLINDNESS = register("blindness", new Potion("blindness", new StatusEffectInstance(StatusEffects.BLINDNESS, 800)));
    public static final RegistryEntry<Potion> LEVITATION = register("levitation", new Potion("levitation", new StatusEffectInstance(StatusEffects.LEVITATION, 1200)));


    public static Item register(String id, Function<Item.Settings, Item> factory, Item.Settings settings) {
        return register(keyOf(id), factory, settings);
    }
    private static RegistryKey<Item> keyOf(String id) {
        return RegistryKey.of(RegistryKeys.ITEM, FixedMinecraft.INSTANCE.id(id));
    }
    public static Item register(RegistryKey<Item> key, Function<Item.Settings, Item> factory, Item.Settings settings) {
        Item item = factory.apply(settings.registryKey(key));
        if (item instanceof BlockItem blockItem) {
            blockItem.appendBlocks(Item.BLOCK_ITEMS, item);
        }

        return Registry.register(Registries.ITEM, key, item);
    }
    public static Item register(Block block) {
        return register(block, BlockItem::new, new Item.Settings());
    }
    public static Item register(Block block, Item.Settings settings) {
        return register(block, BlockItem::new, settings);
    }
    public static Item register(Block block, BiFunction<Block, Item.Settings, Item> factory) {
        return register(block, factory, new Item.Settings());
    }
    public static Item register(Block block, BiFunction<Block, Item.Settings, Item> factory, Item.Settings settings) {
        return register(
                keyOf(block.getRegistryEntry().registryKey()),
                /* method_63751 */ itemSettings -> factory.apply(block, itemSettings),
                settings.useBlockPrefixedTranslationKey()
        );
    }
    private static RegistryKey<Item> keyOf(RegistryKey<Block> blockKey) {
        return RegistryKey.of(RegistryKeys.ITEM, blockKey.getValue());
    }

    private static <T> ComponentType<T> registerComponent(String id, UnaryOperator<ComponentType.Builder<T>> builderOperator) {
        return Registry.register(Registries.DATA_COMPONENT_TYPE, id, builderOperator.apply(ComponentType.builder()).build());
    }

    private static RegistryEntry<Potion> register(String name, Potion potion) {
        return Registry.registerReference(Registries.POTION, FixedMinecraft.INSTANCE.id(name), potion);
    }
}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/ItemRegistry.kt
package net.greenjab.fixedminecraft.registry

import net.greenjab.fixedminecraft.registry.item.EchoFruitItem
import net.greenjab.fixedminecraft.registry.item.TotemItem
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookAdditionsComponent
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookItem
import net.minecraft.component.ComponentType
import net.minecraft.component.type.FoodComponents
import net.minecraft.entity.effect.StatusEffectInstance
import net.minecraft.entity.effect.StatusEffects
import net.minecraft.item.AnimalArmorItem
import net.minecraft.item.FireworkRocketItem
import net.minecraft.item.Item
import net.minecraft.item.equipment.ArmorMaterials
import net.minecraft.potion.Potion
import net.minecraft.registry.Registries
import net.minecraft.registry.Registries.ITEM
import net.minecraft.registry.Registries.POTION
import net.minecraft.registry.Registry
import net.minecraft.sound.SoundEvents
import net.minecraft.util.Rarity
import java.util.function.UnaryOperator


@Suppress("MemberVisibilityCanBePrivate")
object ItemRegistry {
    val DRAGON_FIREWORK_ROCKET = item(::FireworkRocketItem)
    val MAP_BOOK: Item = MapBookItem(Item.Settings().maxCount(1)) as Item
    val MAP_BOOK_ADDITIONS: ComponentType<MapBookAdditionsComponent> = registerComponent("melody_map_book_additions") { builder ->
        builder.codec(
            MapBookAdditionsComponent.CODEC
        ).packetCodec(MapBookAdditionsComponent.PACKET_CODEC).cache()
    }

    private fun <T> registerComponent(id: String, builderOperator: UnaryOperator<ComponentType.Builder<T>>): ComponentType<T> {
        return Registry.register(Registries.DATA_COMPONENT_TYPE, id, builderOperator.apply(ComponentType.builder<T>()).build())
    }


    val NETHERITE_HORSE_ARMOR = item({ AnimalArmorItem(ArmorMaterials.NETHERITE, AnimalArmorItem.Type.EQUESTRIAN, SoundEvents.ENTITY_HORSE_ARMOR, false, it) }) {
        maxCount(1)
        rarity(Rarity.RARE)
        fireproof()
    }

    val BROKEN_TOTEM = item(::Item) {maxCount(1).rarity(Rarity.UNCOMMON)}
     val ECHO_TOTEM: Item = TotemItem(Item.Settings().maxCount(1).rarity(Rarity.UNCOMMON).useCooldown(1.0F))
    // val ECHO_TOTEM: Item = item(::Item) {maxCount(1).rarity(Rarity.UNCOMMON)}
    val ECHO_FRUIT: Item =  EchoFruitItem(Item.Settings().maxCount(64).rarity(Rarity.UNCOMMON).food(FoodComponents.CHORUS_FRUIT))

    val NETHERITE_ANVIL = blockItem(BlockRegistry.NETHERITE_ANVIL, Item.Settings::fireproof)
    val CHIPPED_NETHERITE_ANVIL = blockItem(BlockRegistry.CHIPPED_NETHERITE_ANVIL, Item.Settings::fireproof)
    val DAMAGED_NETHERITE_ANVIL = blockItem(BlockRegistry.DAMAGED_NETHERITE_ANVIL, Item.Settings::fireproof)

    val COPPER_RAIL = blockItem(BlockRegistry.COPPER_RAIL)
    val EXPOSED_COPPER_RAIL = blockItem(BlockRegistry.EXPOSED_COPPER_RAIL)
    val WEATHERED_COPPER_RAIL = blockItem(BlockRegistry.WEATHERED_COPPER_RAIL)
    val OXIDIZED_COPPER_RAIL = blockItem(BlockRegistry.OXIDIZED_COPPER_RAIL)

    val WAXED_COPPER_RAIL = blockItem(BlockRegistry.WAXED_COPPER_RAIL)
    val WAXED_EXPOSED_COPPER_RAIL = blockItem(BlockRegistry.WAXED_EXPOSED_COPPER_RAIL)
    val WAXED_WEATHERED_COPPER_RAIL = blockItem(BlockRegistry.WAXED_WEATHERED_COPPER_RAIL)
    val WAXED_OXIDIZED_COPPER_RAIL = blockItem(BlockRegistry.WAXED_OXIDIZED_COPPER_RAIL)

    val BLINDNESS = Potion("blindness", StatusEffectInstance(StatusEffects.BLINDNESS, 800))
    val LEVITATION = Potion("levitation", StatusEffectInstance(StatusEffects.LEVITATION, 1200))

    fun register() {
        ITEM.register("dragon_firework_rocket", DRAGON_FIREWORK_ROCKET)
        ITEM.register("map_book", MAP_BOOK)
        ITEM.register("netherite_horse_armor", NETHERITE_HORSE_ARMOR)

        ITEM.register("broken_totem", BROKEN_TOTEM)
        ITEM.register("echo_totem", ECHO_TOTEM)
        ITEM.register("echo_fruit", ECHO_FRUIT)

        ITEM.register("netherite_anvil", NETHERITE_ANVIL)
        ITEM.register("chipped_netherite_anvil", CHIPPED_NETHERITE_ANVIL)
        ITEM.register("damaged_netherite_anvil", DAMAGED_NETHERITE_ANVIL)

        ITEM.register("copper_rail", COPPER_RAIL)
        ITEM.register("exposed_copper_rail", EXPOSED_COPPER_RAIL)
        ITEM.register("weathered_copper_rail", WEATHERED_COPPER_RAIL)
        ITEM.register("oxidized_copper_rail", OXIDIZED_COPPER_RAIL)
        ITEM.register("waxed_copper_rail", WAXED_COPPER_RAIL)
        ITEM.register("waxed_exposed_copper_rail", WAXED_EXPOSED_COPPER_RAIL)
        ITEM.register("waxed_weathered_copper_rail", WAXED_WEATHERED_COPPER_RAIL)
        ITEM.register("waxed_oxidized_copper_rail", WAXED_OXIDIZED_COPPER_RAIL)

        POTION.register("blindness", BLINDNESS)
        POTION.register("levitation", LEVITATION)
    }
}
